### DISCUSSION TITLE
What is the difference between passing the statement Chem.MolFromSmiles directly or via a variable?

### DISCUSSION THREAD

**Discussion by Gluroda** [rdkit/rdkit#5925] [Q&A] [answered]
If I do not store the `rdkit.Chem.rdchem.Mol` object in a variable but pass the statement `Chem.MolFromSmiles("<your-smile>")` directly into another function it gives a different result than storing it in a variable before!

Why is that?

```
>>> from rdkit.Chem import Descriptors
>>> from rdkit import Chem



>>> # direct approach
>>> print(Descriptors.TPSA(Chem.MolFromSmiles('OC(=O)P(=O)(O)O')))
94.83
>>> print(Descriptors.TPSA(Chem.MolFromSmiles('OC(=O)P(=O)(O)O'), includeSandP=True))
104.64000000000001



>>> # mol as variable approach
>>> mol = Chem.MolFromSmiles('OC(=O)P(=O)(O)O')
>>> print(Descriptors.TPSA(mol))
94.83
>>> print(Descriptors.TPSA(mol, includeSandP=True))
94.83
```

In my mind the last `print`statement should also give a result of \~104.64

This links you to the example that I am using: [TPSA][1]

  [1]: https://rdkit.org/docs/RDKit_Book.html?highlight=tpsa#implementation-of-the-tpsa-descriptor

**Comment by greglandrum** [comment, marked as answer]
@Gluroda when you call the TPSA function on a molecule, the results of calling it are cached on the molecule itself (so if you call it a second time you get the results much faster). In this case if you want different parameters you have to force the re-calculation:
```
In [3]: mol = Chem.MolFromSmiles('OC(=O)P(=O)(O)O')

In [4]: print(Descriptors.TPSA(mol,includeSandP=True))
104.64000000000001

In [5]: print(Descriptors.TPSA(mol,includeSandP=False))
104.64000000000001

In [7]: print(Descriptors.TPSA(mol,force=True,includeSandP=False))
94.83
```

This is actually a bug though: the caching should be smart enough to recognize that you've called the function with different arguments and adjust accordingly.

**Reply by Gluroda** [reply]
Hey, thanks for your explanation. 

I understand what is going "wrong" there and agree with you that it is kind of inconvenient for users that do not know about the caching location. 
Does this also affect other functions? I am kind of scared now...

I am grateful that you provided a "workaround" via `force=True` and will use it for this special case.

Should I report it as a bug anywhere? I would be happy to be of any help.

**Comment by bp-kelley** [comment]
I consider this a bug.  We should cache the include S and P = True differently from False.

**Reply by greglandrum** [reply]
> I consider this a bug. We should cache the include S and P = True differently from False.

Agreed... that's what I said at the end of my message.

**Comment by bp-kelley** [comment]
So it is :)
